### ISSUE TITLE
Do not report MA0004 for an IAsyncDisposable?

### ISSUE THREAD

**Issue by drieseng** [meziantou/Meziantou.Analyzer#619]
**MA0004** (and **CA2007**) reports a diagnostic for types that implement **IAsyncDisposable**.
This, together with **MA0042** (prefer using "await using"), results in unusable code.

Let's start with this:
```cs
using System.IO;

 using (var fs = File.Open("whatever"))
{
    await file.FlushAsync().ConfigureAwait(false);
}
```

Since **Stream** implements **IAsyncDisposable**, **MA0042** will kick in and ask you to write the code fragment as follows:

```cs
using System.IO;

await using (var fs = File.Open("whatever"))
{
    await file.FlushAsync().ConfigureAwait(false);
}
```

All good, but now **MA0004** (and **CA2007**) wake up and ask you to add a `ConfigureAwait(false)`:

```cs
using System.IO;

await using (var fs = File.Open("whatever").ConfigureAwait(false))
{
    await file.FlushAsync().ConfigureAwait(false);
}
```

Problem is that for a type that implements **IAsyncDisposable**, `ConfigureAwait(false)` is implemented as an extension method. That extension method returns a **ConfiguredAsyncDisposable**, not a **ConfiguredTaskAwaitable\<TResult>**:

```cs
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace System.Threading.Tasks
{
    /// <summary>Provides a set of static methods for configuring <see cref="Task"/>-related behaviors on asynchronous enumerables and disposables.</summary>
    public static partial class TaskAsyncEnumerableExtensions
    {
        /// <summary>Configures how awaits on the tasks returned from an async disposable will be performed.</summary>
        /// <param name="source">The source async disposable.</param>
        /// <param name="continueOnCapturedContext">Whether to capture and marshal back to the current context.</param>
        /// <returns>The configured async disposable.</returns>
        public static ConfiguredAsyncDisposable ConfigureAwait(this IAsyncDisposable source, bool continueOnCapturedContext) =>
            new ConfiguredAsyncDisposable(source, continueOnCapturedContext);

        ...
    }
}
```

The result is that you end up losing (typed) access to what you awaited. The code fragment I provided no longer compiles as `FlushAsync()` does not exist on **ConfiguredAsyncDisposable**.

**Comment by meziantou** [comment]
The fixed code should be

```
// not elegant, but it's the only way to do it
var fs = File.Open("whatever");
await using (fs.ConfigureAwait(false))
{
}
```
